### PR TITLE
McChainSimulator (major) patch

### PR DESCRIPTION
**Pull request by ramirezdiego**
**What is the problem / what does the code in this PR do**
Hunting some bugs in the McChainSimulator class, to facilitate our life on the epix and mc_chain side.

For the moment I just blindly pushed a fix discussed with @petergaemers, since we get `AttributeError: 'McChainSimulator' object has no attribute 'instructions_nveto'` when running the code with nVeto set to `st.set_config(dict(neutron_veto=False))`.

(To be continued)

### CODE REVIEW

**Comment by ramirezdiego** [comment]
Folowing the discussion in https://github.com/XENONnT/WFSim/issues/126, I removed the `set_max` and `instructions_from_epix_cut` functions in the class.

**Comment by ramirezdiego** [comment]
@zhut19 Could you please have a quick look at the last commit only? I tried to align this branch with master, after you merged #117, but using the new syntax for the config IF statements.

**Comment by zhut19** [comment]
> @zhut19 Could you please have a quick look at the last commit only? I tried to align this branch with master, after you merged #117, but using the new syntax for the config IF statements.

Looks good to me :)

**Comment by ershockley** [comment]
Hey guys, is there an ETA on when this might be merged? I'd like to test some jobs out on the grid but I understand the mc chain needs this branch. If it will be more than a ~week to merge I'd like to know so I can setup a test environment.

**Comment by zhut19** [comment]
> Hey guys, is there an ETA on when this might be merged? I'd like to test some jobs out on the grid but I understand the mc chain needs this branch. If it will be more than a ~week to merge I'd like to know so I can setup a test environment.

I am still working on some code clean up and testing it out but should be available for further review in a day or two.

File: wfsim/load_resource.py [hunk 1]
```diff
@@ -114,26 +114,25 @@ def __init__(self, config=None):
...
+            if config.get('enable_gas_gap_warping',False):
                 self.gas_gap_length = make_map(["constant dummy", 0.25, [254,]])
 
             # Photon After Pulses
-            if config['enable_pmt_afterpulses']:
+            if config.get('enable_pmt_afterpulses',False):
                 self.uniform_to_pmt_ap = straxen.get_resource(files['photon_ap_cdfs'], fmt='pkl.gz')
 
-        if config['detector'] == 'XENONnT':
+        if config.get('detector','XENONnT') == 'XENONnT' and not config.get('neutron_veto', False):
```
Comment (ramirezdiego) [inline wfsim/load_resource.py:135]: @mzks Why is this change necessary?

Comment (mzks) [reply wfsim/load_resource.py:135]: nVeto side doesn't use the following files, thus we can skip them for nVeto.
If we don't add them, it works.

Comment (ramirezdiego) [reply wfsim/load_resource.py:135]: Ok, for a moment I was happily thinking that WFSim runs simultaneously for TPC and nVeto, while every time the `neutron_veto` config changes, it actually means we updated our context and started a new WFSim run :)

File: wfsim/strax_interface.py [hunk 2]
```diff
@@ -266,8 +248,8 @@ def set_timing(self,):
     
     def set_configuration(self,):
         """Set chunking configuration and feeds instructions to wfsim"""
-        max_events_per_chunk=500#is this hard coding really nessecairy...
-        chunk_size = np.ceil(max_events_per_chunk/self.context.config['event_rate'])
+        max_events_per_chunk=20 #is this hard coding really nessecairy...
```
Comment (ramirezdiego) [inline wfsim/strax_interface.py:251]: @petergaemers Can you help me understand if this can become problematic for very high rates?

**Comment by petergaemers** [comment]
I included a nveto config override and changed the references to "provides" to "targets" which is a config option

File: wfsim/strax_interface.py [hunk 3]
```diff
@@ -632,6 +632,7 @@ def set_timing(self,):
             self.instructions_epix = self.instructions_epix[~extra_long]
             log.warning('Found and removing %d epix instructions later than maximum time %d'
                         % (extra_long.sum(), max_time))
+            self.instrutions_epix = np.sort(self.instructions_epix, order='time')
```
Comment (ramirezdiego) [inline wfsim/strax_interface.py:635]: We need to add the same for the nVeto target, don't we?

File: wfsim/strax_interface.py [hunk 4]
```diff
@@ -621,147 +480,285 @@ def _setup(self):
...
+        # nveto instruction doesn't carry physical time delay, so the time is overwritten
+        if 'nveto' in self.config['targets']:
+            i_timings = np.searchsorted(np.arange(self.config['entry_start'], self.config['entry_stop']),
+                                        self.instructions_nveto['g4id'])
+            self.instructions_nveto['time'] += timings[i_timings]
+
+            extra_long = self.instructions_nveto['time'] > max_time
+            self.instructions_nveto = self.instructions_nveto[~extra_long]
+            log.warning('Found and removing %d nveto instructions later than maximum time %d'
+                        % (extra_long.sum(), max_time))
```
Comment (ramirezdiego) [inline wfsim/strax_interface.py:646]: Here, and in the TPC case a few lines above, the `max_time` reference is the G4 event zero time. We need to discuss this with actual examples, as valid events can arise that simply take longer to produce the first interaction wrt this zero time.

**Comment by petergaemers** [comment]
Electron afterpulses still lead to a crash. Although it's a different one than before
https://gist.github.com/petergaemers/91b6f2002a1610d53638fd718abd12a7

File: wfsim/strax_interface.py (hunk 3 above)
Comment (zhut19) [reply wfsim/strax_interface.py:635]: But there's the sorting done in the https://github.com/XENONnT/WFSim/blob/master/wfsim/core.py#L1163-L1164 anyway. Sorting here is not necessary.

**Comment by zhut19** [comment]
> Electron afterpulses still lead to a crash. Although it's a different one than before
> https://gist.github.com/petergaemers/91b6f2002a1610d53638fd718abd12a7

This looks like the instruction buffer run out of space. Although I don't know why this is happening, there's really this many afterpulses? Anyway, we should be able to get away with removing a few electron afterpulses, so I did just that in https://github.com/XENONnT/WFSim/pull/125/commits/40a7f73c13a55214ba2eb4520e7562f92fb3c6e4

File: wfsim/strax_interface.py (hunk 4 above)
Comment (zhut19) [reply wfsim/strax_interface.py:643]: Both the `self.instructions_nveto['time']` and `max_time` are absolute time at this point. The instructions removed here are cases that the time of the instruction is so late that we can not fit them into this batch, (max time defined by `entry_stop / event_rate`

File: wfsim/strax_interface.py (hunk 3 above)
Comment (ramirezdiego) [reply wfsim/strax_interface.py:632]: Okay, this might be the reason for the misunderstanding with @petergaemers that we thought we had resolved this morning. Then the instructions (even from epix are always sorted?

File: wfsim/strax_interface.py (hunk 4 above)
Comment (ramirezdiego) [reply wfsim/strax_interface.py:643]: Thanks @zhut19, now I see `max_time = int((self.config['entry_stop'] + 0.5) / rate)` a few lines above. Two comments then:

1. With that definition, doesn't `max_time` risk to be identical than the starting time of the last entry? Shouldn't we add a buffer to `max_time`?
2. We still need to discuss what to do with the late G4 entries (we apply a cut in epix to entries 10ms after the first one in in LXe (so not in absolute event time), but the same isn't done in the nVeto side.

Comment (zhut19) [reply wfsim/strax_interface.py:643]: 1. 
    - Depending on our event rate, even with a typical delay between the signal and the G4 time of 1ms, and an event rate of 100 Hz, we would have 0.1 events outside the max time per batch.
    -  Adding a buffer after max_time doesn't solve the issue completely since there is always this delay (between the signal and the G4 time), but can very likely create overlapping with the next batch.
2.
    - The logic implemented here is a bit different, thus I am not sure any 10 ms cut-off is necessary. 
    - Here, for example, uniform timing is sampled for each G4id and added to all signal timings to make this combined timing, we only remove the really delayed signals that happen after the max time. In which case for the first entry, the delay time cut-off is ~ 10 seconds (100 Hz, 1000 evt/batch), but for the last entry, the delay time cut-off is ~ 10 milliseconds (100 Hz).
    - For nveto, the newly added function in https://github.com/XENONnT/WFSim/pull/125/commits/99572a8c4b7845b021c643846b08a5db6cb59c0a ensures that each instruction time is actually the signal time with the photons of each instruction happen at the latest 1ms afterward. And in the set timing step for nveto, the same time combining and max time cut-off logic is applied.

**Review by petergaemers — APPROVED** [review]

File: wfsim/strax_interface.py (hunk 3 above)
Comment (petergaemers) [reply wfsim/strax_interface.py]: Good catch, so we can remove this

File: wfsim/strax_interface.py (hunk 4 above)
Comment (ramirezdiego) [reply wfsim/strax_interface.py:626]: >     1. * Depending on our event rate, even with a typical delay between the signal and the G4 time of 1ms, and an event rate of 100 Hz, we would have 0.1 events outside the max time per batch.
>        * Adding a buffer after max_time doesn't solve the issue completely since there is always this delay (between the signal and the G4 time), but can very likely create overlapping with the next batch.
>

Just to confirm, here you say "next batch" as referring to future stages in which we merge various WFSim outputs into superruns, correct?
 
>     2. * The logic implemented here is a bit different, thus I am not sure any 10 ms cut-off is necessary.
>        * Here, for example, uniform timing is sampled for each G4id and added to all signal timings to make this combined timing, we only remove the really delayed signals that happen after the max time. In which case for the first entry, the delay time cut-off is ~ 10 seconds (100 Hz, 1000 evt/batch), but for the last entry, the delay time cut-off is ~ 10 milliseconds (100 Hz).

Okay, in this case I will make sure to test a few different type of simulations and see whether we remove lots of epix delayed instructions (there's now a printout for this thanks to you!). But I agree that the epix cutoff doesn't need to be matched with nVeto a priori (will keep thinking about this...).

>        * For nveto, the newly added function in [99572a8](https://github.com/XENONnT/WFSim/commit/99572a8c4b7845b021c643846b08a5db6cb59c0a) ensures that each instruction time is actually the signal time with the photons of each instruction happen at the latest 1ms afterward. And in the set timing step for nveto, the same time combining and max time cut-off logic is applied.

👍

**Comment by PavelKavrigin** [comment]
I see a new error after the latest commit. These are the branches that I use for tests:

mc_chain: 'wfsim' branch
WFSim: 'fullchain_fixes' branch
private_nt_aux_files: 'nveto_wfsim_config' branch

Here's the full output:
<details>
  <output>

```
Generate raw_records for job ib 1
epix configuration:  {'cut_by_eventid': True, 'debug': True, 'source_rate': 0, 'micro_separation_time': 10.0, 'max_delay': 10000000.0, 'input_file': '/home/pavelk/tmp/tpc_and_nveto_cryoneutrons/tpc_and_nveto_cryoneutrons_200.root', 'detector': 'XENONnT', 'detector_config_override': False, 'micro_separation': 0.05, 'tag_cluster_by': 'time', 'entry_start': 0, 'entry_stop': 100, 'path': '/home/pavelk/tmp/tpc_and_nveto_cryoneutrons', 'file_name': 'tpc_and_nveto_cryoneutrons_200.root', 'detector_config': [<epix.detector_volumes.SensitiveVolume object at 0x7f537e66ac70>, <epix.detector_volumes.SensitiveVolume object at 0x7f537e6ea160>], 'outer_cylinder': {'max_z': 7.3936, 'min_z': -154.6555, 'max_r': 66.4}}
Total entries in input file = 198
Starting to read from g4 eventid 0
Ending read in at g4 eventid 100
It took 3.7585 sec to load data.
Finding clusters of interactions with a dr = 0.05 mm and dt = 10.0 ns
It took 4.9347 sec to find clusters.
It took 4.2800 sec to merge clusters.
Removing clusters not in volumes: TPC BelowCathode
Number of clusters before: 1430
Number of clusters after: 1426
Assigning electric field to clusters
Generating photons and electrons for events
/home/pavelk/notebooks/mc/epix/quanta_generation.py:51: UserWarning: Energy deposition of 263.45294284820557 keV beyond NEST validity for NR model of 200 keV - Remove Interaction
  warnings.warn(f"Energy deposition of {en} keV beyond NEST validity for NR model of 200 keV - Remove Interaction")
/home/pavelk/notebooks/mc/epix/quanta_generation.py:51: UserWarning: Energy deposition of 216.469929844141 keV beyond NEST validity for NR model of 200 keV - Remove Interaction
  warnings.warn(f"Energy deposition of {en} keV beyond NEST validity for NR model of 200 keV - Remove Interaction")
/home/pavelk/notebooks/mc/epix/quanta_generation.py:51: UserWarning: Energy deposition of 228.91101399809122 keV beyond NEST validity for NR model of 200 keV - Remove Interaction
  warnings.warn(f"Energy deposition of {en} keV beyond NEST validity for NR model of 200 keV - Remove Interaction")
/home/pavelk/notebooks/mc/epix/quanta_generation.py:51: UserWarning: Energy deposition of 431.4281909974525 keV beyond NEST validity for NR model of 200 keV - Remove Interaction
  warnings.warn(f"Energy deposition of {en} keV beyond NEST validity for NR model of 200 keV - Remove Interaction")
/home/pavelk/notebooks/mc/epix/quanta_generation.py:51: UserWarning: Energy deposition of 272.36525202402845 keV beyond NEST validity for NR model of 200 keV - Remove Interaction
  warnings.warn(f"Energy deposition of {en} keV beyond NEST validity for NR model of 200 keV - Remove Interaction")
/home/pavelk/notebooks/mc/epix/quanta_generation.py:51: UserWarning: Energy deposition of 328.93114709854126 keV beyond NEST validity for NR model of 200 keV - Remove Interaction
  warnings.warn(f"Energy deposition of {en} keV beyond NEST validity for NR model of 200 keV - Remove Interaction")
/home/pavelk/notebooks/mc/epix/quanta_generation.py:51: UserWarning: Energy deposition of 294.90637615241576 keV beyond NEST validity for NR model of 200 keV - Remove Interaction
  warnings.warn(f"Energy deposition of {en} keV beyond NEST validity for NR model of 200 keV - Remove Interaction")
/home/pavelk/notebooks/mc/epix/quanta_generation.py:51: UserWarning: Energy deposition of 463.1151755563915 keV beyond NEST validity for NR model of 200 keV - Remove Interaction
  warnings.warn(f"Energy deposition of {en} keV beyond NEST validity for NR model of 200 keV - Remove Interaction")
/home/pavelk/notebooks/mc/epix/quanta_generation.py:51: UserWarning: Energy deposition of 328.00815400481224 keV beyond NEST validity for NR model of 200 keV - Remove Interaction
  warnings.warn(f"Energy deposition of {en} keV beyond NEST validity for NR model of 200 keV - Remove Interaction")
/home/pavelk/notebooks/mc/epix/quanta_generation.py:51: UserWarning: Energy deposition of 391.00357958767563 keV beyond NEST validity for NR model of 200 keV - Remove Interaction
  warnings.warn(f"Energy deposition of {en} keV beyond NEST validity for NR model of 200 keV - Remove Interaction")
/home/pavelk/notebooks/mc/epix/quanta_generation.py:51: UserWarning: Energy deposition of 363.09867829084396 keV beyond NEST validity for NR model of 200 keV - Remove Interaction
  warnings.warn(f"Energy deposition of {en} keV beyond NEST validity for NR model of 200 keV - Remove Interaction")
/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.8/site-packages/numpy/lib/function_base.py:2192: RuntimeWarning: invalid value encountered in ? (vectorized)
  outputs = ufunc(*inputs)
It took 1.8180 sec to get quanta.
WARNING:wfsim.resource:Trying to use the private repo to load XENONnT_spe_distributions_nveto_013071.csv locally
WARNING:wfsim.resource:photon_area_distribution did not download, trying /home/pavelk/private_nt_aux_files/sim_files/XENONnT_spe_distributions_nveto_013071.csv
WARNING:wfsim.resource:Trying to use the private repo to load xnt_noise_nveto_014901.npz locally
WARNING:wfsim.resource:noise_file did not download, trying /home/pavelk/private_nt_aux_files/sim_files/xnt_noise_nveto_014901.npz
WARNING:wfsim.interface:Found and removing 0 epix instructions later than maximum time 100500000000
WARNING:wfsim.interface:Found and removing 0 nveto instructions later than maximum time 100500000000
WARNING:wfsim.resource:Trying to use the private repo to load XENONnT_GARFIELD_B1d5n_C30n_G1n_A6d5p_T1d5n_PMTs1d5n_FSR0d95n.npz locally
WARNING:wfsim.resource:s2_luminescence did not download, trying /home/pavelk/private_nt_aux_files/sim_files/XENONnT_GARFIELD_B1d5n_C30n_G1n_A6d5p_T1d5n_PMTs1d5n_FSR0d95n.npz
WARNING:wfsim.resource:Trying to use the private repo to load  locally
WARNING:wfsim.resource:field_dependencies_map did not download, trying /home/pavelk/private_nt_aux_files/sim_files/
WARNING:strax:Could not estimate run start and end time from run metadata: assuming it is 0 and inf
Simulating Raw Records:   0%|          | 0/56 [00:00<?, ?it/s]Exception in thread CRITICAL:ThreadedMailboxProcessor:Target Mailbox (raw_records) killed, exception <class 'strax.mailbox.MailboxKilled'>, message (<class 'ValueError'>, ValueError('shape mismatch: value array of shape (0,) could not be broadcast to indexing result of shape (4,)'), <traceback object at 0x7f54a89e3300>)
divide_outputs:raw_records:
Traceback (most recent call last):
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.8/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.8/site-packages/strax/mailbox.py", line 267, in _send_from
    self.kill_from_exception(e)
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.8/site-packages/strax/mailbox.py", line 186, in kill_from_exception
    raise e
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.8/site-packages/strax/mailbox.py", line 254, in _send_from
    x = next(iterable)
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.8/site-packages/strax/plugin.py", line 394, in iter
    yield self.do_compute(chunk_i=chunk_i, **inputs_merged)
  File "/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.8/site-packages/strax/plugin.py", line 492, in do_compute
    result = self.compute(**kwargs)
  File "/home/pavelk/notebooks/utils/wfsim/strax_interface.py", line 689, in compute
    result = next(self.sim_iter)
  File "/home/pavelk/notebooks/utils/wfsim/strax_interface.py", line 229, in __call__
    for channel, left, right, data in self.rawdata(instructions=instructions,
  File "/home/pavelk/notebooks/utils/wfsim/core.py", line 1258, in __call__
    instb[ib] = instb_secondary[:-n_too_much]
ValueError: shape mismatch: value array of shape (0,) could not be broadcast to indexing result of shape (4,)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-7-82839ced3a45> in <module>
      9 
     10     print(f"Generate raw_records for job ib {args.run_id}")
---> 11     st.make(args.run_id, 'raw_records')
     12 
     13     if args.neutron_veto:

/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.8/site-packages/strax/context.py in make(self, run_id, targets, save, max_workers, progress_bar, _skip_if_built, **kwargs)
   1119             return
   1120 
-> 1121         for _ in self.get_iter(run_ids[0], targets,
   1122                                progress_bar=progress_bar,
   1123                                save=save, max_workers=max_workers, **kwargs):

/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.8/site-packages/strax/context.py in get_iter(self, run_id, targets, save, max_workers, time_range, seconds_range, time_within, time_selection, selection_str, keep_columns, allow_multiple, progress_bar, _chunk_number, **kwargs)
    979 
    980         except Exception as e:
--> 981             generator.throw(e)
    982             raise
    983 

/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.8/site-packages/strax/context.py in get_iter(self, run_id, targets, save, max_workers, time_range, seconds_range, time_within, time_selection, selection_str, keep_columns, allow_multiple, progress_bar, _chunk_number, **kwargs)
    956             with _p as pbar:
    957                 pbar.last_print_t = time.time()
--> 958                 for n_chunks, result in enumerate(strax.continuity_check(generator), 1):
    959                     seen_a_chunk = True
    960                     if not isinstance(result, strax.Chunk):

/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.8/site-packages/strax/chunk.py in continuity_check(chunk_iter)
    264     last_end = None
    265     last_runid = None
--> 266     for s in chunk_iter:
    267         if s.run_id != last_runid:
    268             # TODO: can we do better?

/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.8/site-packages/strax/processor.py in iter(self)
    268             # which is printed for the user
    269             self.log.debug("Reraising exception")
--> 270             raise exc.with_traceback(traceback)
    271 
    272         # Check the savers for any exception that occurred during saving

/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.8/site-packages/strax/mailbox.py in _send_from(self, iterable)
    252 
    253                 try:
--> 254                     x = next(iterable)
    255                 except StopIteration:
    256                     # No need to send this yet, close will do that

/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.8/site-packages/strax/plugin.py in iter(self, iters, executor)
    392                     yield new_future
    393                 else:
--> 394                     yield self.do_compute(chunk_i=chunk_i, **inputs_merged)
    395 
    396         except IterDone:

/opt/XENONnT/anaconda/envs/XENONnT_development/lib/python3.8/site-packages/strax/plugin.py in do_compute(self, chunk_i, **kwargs)
    490             kwargs['start'] = start
    491             kwargs['end'] = end
--> 492         result = self.compute(**kwargs)
    493 
    494         return self._fix_output(result, start, end)

~/notebooks/utils/wfsim/strax_interface.py in compute(self)
    687         if 'tpc' in self.config['targets']:
    688             try:
--> 689                 result = next(self.sim_iter)
    690             except StopIteration:
    691                 if self.sim.source_finished():

~/notebooks/utils/wfsim/strax_interface.py in __call__(self, instructions, time_zero, **kwargs)
    227         self.chunk_time = self.chunk_time_pre + cksz  # Starting chunk
    228         self.current_digitized_right = self.last_digitized_right = 0
--> 229         for channel, left, right, data in self.rawdata(instructions=instructions,
    230                                                        truth_buffer=self.truth_buffer,
    231                                                        **kwargs):

~/notebooks/utils/wfsim/core.py in __call__(self, instructions, truth_buffer, progress_bar, **kwargs)
   1256                                 log.warning(f'Running out of instruction buffer removing {n_too_much} secondaries')
   1257                             ib = np.where(~instb_filled)[0][:len(instb_secondary) - n_too_much]
-> 1258                             instb[ib] = instb_secondary[:-n_too_much]
   1259                             instb_filled[ib] = True
   1260 

ValueError: shape mismatch: value array of shape (0,) could not be broadcast to indexing result of shape (4,)
```    
  </output>  
</details>

**Comment by zhut19** [comment]
> I see a new error after the latest commit. These are the branches that I use for tests:

Thanks! And sorry about the bug, should have seen this one coming.

**Comment by petergaemers** [comment]
Merge time?

**Comment by zhut19** [comment]
> One question for now: this new branch would work In combination with [XENONnT/mc_chain#14](https://github.com/XENONnT/mc_chain/pull/14). in which for the moment we have disabled `enable_electron_afterpulses` (known issue). I also saw that `check_raw_record_overlaps` is also disabled for TPC records. Is this something we aim at solving here?

You can try enabling the electron afterpulses, hopefully, the issue is resolved. Somehow, there are overlapping pulses at the moment, but the code can run with this check raw record overlaps set to faluse, I am not aiming to solve it now.

**Comment by PavelKavrigin** [comment]
I see a new error after the latest commit (or maybe because of the new Straxen release?) Please see the full output in the attachment. The error is "AssertionError: Run out of instruction buffer". It is also alarming that we see this warning before the error: "WARNING:wfsim.interface:Found and removing 22 epix instructions later than maximum time". I am using the same branches as mentioned above, I've pulled all updates this morning (2021-04-18).   
  [wfsim_err.txt](https://github.com/XENONnT/WFSim/files/6331307/wfsim_err.txt)

**Comment by ramirezdiego** [comment]
@PavelKavrigin I can't comment on the buffer error, but I don't think it has anything to do with the cut on the epix instructions, which was already present before (and now is more visible since Tianyu changed the code from debug to warning mode).

For the moment, these 22 entries lost are the price to pay for the continuous time format. See conversation above: https://github.com/XENONnT/WFSim/pull/125#discussion_r611856381.

**Comment by ramirezdiego** [comment]
Sorry, I'm on the phone and I tapped on the wrong button :)

**Comment by petergaemers** [comment]
The cut events are the super delayed events we keep talking about

**Comment by zhut19** [comment]
> I see a new error after the latest commit (or maybe because of the new Straxen release?) Please see the full output in the attachment. The error is "AssertionError: Run out of instruction buffer".

Hi @PavelKavrigin I don't see this error when I run with the latest commit. But it's possible if the event rate is too high, like > 1000 Hz. Can you give me more detail on how where this error come from? (I mean that conditions you put in the config). Thanks!

**Comment by PavelKavrigin** [comment]
> > I see a new error after the latest commit (or maybe because of the new Straxen release?) Please see the full output in the attachment. The error is "AssertionError: Run out of instruction buffer".
> 
> Hi @PavelKavrigin I don't see this error when I run with the latest commit. But it's possible if the event rate is too high, like > 1000 Hz. Can you give me more detail on how where this error come from? (I mean that conditions you put in the config). Thanks!

I use:

straxen : 0.18.0
uproot : 4.0.7
wfsim : 0.4.1
epix : 0.1.1

mc_chain: 'wfsim' branch
WFSim: 'fullchain_fixes' branch
private_nt_aux_files: 'nveto_wfsim_config' branch

Please find the test code (including the configuration parameters) and the output in the attachment. I use 'tpc_and_nveto_cryoneutrons_200.root' example with event_rate==10.  
[test_error.txt](https://github.com/XENONnT/WFSim/files/6337630/test_error.txt)
[test_code.txt](https://github.com/XENONnT/WFSim/files/6337636/test_code.txt)

**Comment by PavelKavrigin** [comment]
> The cut events are the super delayed events we keep talking about. The buffer size is 10000 instructions, are you simulating something huge giving a lot of afterpulses?

OK, I see. It is 'tpc_and_nveto_cryoneutrons_200.root' example.

**Comment by petergaemers** [comment]
With enable_electron_afterpulses=False this error goes away, so it's related to that guy

**Comment by zhut19** [comment]
> With enable_electron_afterpulses=False this error goes away, so it's related to that guy. Of course, an alternative is to simply increase the buffer size

Okay, that's why I couldn't reproduce this error, I run mine with this also set to False. I don't know how to fix this quickly, now as each afterpulse single electron occupies a line in the instruction buffer, there's always the possibility to fill the buffer. @PavelKavrigin can you run with afterpulse = False for now? we can find a long-term solution in another pr in the future.

**Comment by PavelKavrigin** [comment]
Yes, 'enable_electron_afterpulses=False' solves the issue. I forgot to change it after I had copied the code from the latest version of 'run_wfsim.py'. Thank you!

**Comment by ramirezdiego** [comment]
The code runs happily with the designed full-chain scripts (https://github.com/XENONnT/mc_chain/pull/14) and an event-by-event check shows matching between the original G4 times distribution and the one after WFSim.

As agreed with @PavelKavrigin and @petergaemers, we need to move forward here and start stress-testing in the mc_chain end, but a more systematic check of the correct event time distribution and TPC-nVeto time consistency is necessary and we'll push for that as a dedicated MC task. This systematic check became complicated, as the final events selection might ignore entries in which the S2 is simply too small, due to the known SE gain issue (https://github.com/XENONnT/WFSim/issues/120).
